### DISCUSSION TITLE
Make custom interfaces first class citizens

### DISCUSSION THREAD

**Discussion by Kematia** [directus/directus#5386] [Feature Requests]
Hey gang! So, I've been working a bit on more advanced custom extensions, and while my experience overall has been good, I've came across some pain points. I had seen some discussion floating around already but I can't find it anymore. 

The gist of it was making custom extensions "first class citizens". Directus is an absolutely amazing project and I feel like a solid custom extensions DX is one of the core attractions to its target users 😄 

## Expose the directus app/api internals to custom interfaces

Working off the NPM package you don't have access (or atleast, I didn't find a way) to internal APIs (composables, utils), private components (drawer-collection), etcetera.

For example, without access to drawer-collection I had to re-implement a lot of logic (search, filters) which was absent from the exposed v-drawer component.

One solution I tried was to have a dummy directus source folder to import the functions, which would work when developing but I couldn't get Rollup to build them correctly.

One solution could be to publish an optional package (_directus-dev?_) with just the source code you could import and build, a companion to the usual directus package.

Or you could include the `src` folder in the npm package?

## Drop in already existing components
This would also enable us to _drop in_ existing directus/app components and tweak them to our liking. As it is right now, my layout-table and directus layout-table script looks vastly different.

## Hot reloading
This _directus-dev_ package could also expose a dev environment, similar to `npm run dev` of the source application. You can get something similar with tools like nodemon, but the DX is a bit worse, as you have to reload the page to see your changes. Which is fine if you're working on page-level interfaces, but it gets a little more tiring on stuff like nested drawers.

_Pro tip! If you have the directus source running with `npm run dev` you can copyfiles or rollup your interfaces to the `app/extensions/_custom` folder and they will hot reload!_

## Buildless Custom Interfaces
Ok, this one may be a bit tricky! The biggest pain point was working with Rollup, which ~~is a undocumented piece of~~ has bad developer ergonomics. Getting TS to work, resolving modules and importing the directus core components was... frustrating to say the least. Having the possibility of skipping Rollup entirely would make developing custom extensions much more attractive.

***

### Ad Interim
![image](https://user-images.githubusercontent.com/17851386/116674698-44914080-a9a5-11eb-863c-a0ab2b27bc34.png)

After spending a couple days banging my head against all of the above, I decided to drop trying to work with rollup and the npm package. I will probably publish my own fork/package* with my custom interfaces scoped to a `_custom` folder, just so I don't have to worry too much when merging in the directus updates. 
I really don't want to go back to nodemon/rollup after working with `npm run dev` hot reloading 😄 

Looking forward to your feedback!

***
*Licence permitting

**Comment by rijkvanzanten** [comment]
> Expose the directus app/api internals to custom interfaces

This could be solved by starting to put things into a `@directus/shared` (name tbd) package that makes it a little easier to import those sorts of available internals, without having to rely on provide/inject for _everything_

> Drop in already existing components

Not sure if this is necessarily directly related, as most components are already globally registered in the Vue context 🤔 Would having them available in a package improve anything on this front?

> Hot reloading

Still wondering if we should come up with a sort of "Directus Storybook" that renders a minimal UI wrapper that allows you to quickly test interface options, or if we should focus our efforts on locally running "regular prod Directus" and then have the interfaces auto-reload (cc @Jonesus)

> Buildless Custom Interfaces

This one is gonna be a little tricky. @nickrum and I have been discussing a sort of included bundler in _runtime_, but I think it's safer to rely on a pre-made extensions toolkit that you can use, so you don't have to mess around with rollup yourself (like @Jonesus already setup in https://github.com/directus-community/extension-toolkit)

**Reply by Kematia** [reply]
> > Expose the directus app/api internals to custom interfaces
> 
> This could be solved by starting to put things into a `@directus/shared` (name tbd) package that makes it a little easier to import those sorts of available internals, without having to rely on provide/inject for _everything_

That would work! Seems like a good middle ground. The key point is being able to use the `import x from @directus/shared` to get IDE goodies. Then the bundling could be at runtime or buildtime.

> > Drop in already existing components
> 
> Not sure if this is necessarily directly related, as most components are already globally registered in the Vue context 🤔 Would having them available in a package improve anything on this front?

This is a much more niche use case, but there are some cases where the directus components are _almost_ perfect, but just need a little tweaking. 

Kinda like what I did for the layout filters for the tabular layout. In that case I had to rewrite a lot of code to get the custom tabular component to work like the directus component, seeing one had access to the internals via `import`, and was using ts and the composition-api, while the other didn't have access to the internals, and was using vanilla vue options-api. If both used the same build 'chain', I could just copy and paste the current directus component and have it work from the get go. 

Another time-saver would be being able to import the scss mixins like `form-grid`...
 
> > Hot reloading
> 
> Still wondering if we should come up with a sort of "Directus Storybook" that renders a minimal UI wrapper that allows you to quickly test interface options, or if we should focus our efforts on locally running "regular prod Directus" and then have the interfaces auto-reload (cc @Jonesus)

I'd think that the closest we can get to the source package `npm run dev`, the less headaches we might get. But this is a bit above my paygrade so I'll let you guys think of the easiest way to implement it. It is a lifesaver for developing custom extensions tho!

> > Buildless Custom Interfaces
> 
> This one is gonna be a little tricky. @nickrum and I have been discussing a sort of included bundler in _runtime_, but I think it's safer to rely on a pre-made extensions toolkit that you can use, so you don't have to mess around with rollup yourself (like @Jonesus already setup in https://github.com/directus-community/extension-toolkit)

I do think that the community could come up with a toolkit. This might be easier with the directus/shared package? Even tho most of the errors I was getting where coming from trying to integrate typescript.

**Reply by Jonesus** [reply]
> Still wondering if we should come up with a sort of "Directus Storybook" that renders a minimal UI wrapper that allows you to quickly test interface options, or if we should focus our efforts on locally running "regular prod Directus" and then have the interfaces auto-reload (cc @Jonesus)

I think the holy grail of any development is that the UI reloads itself on any code changes (maybe even persisting current state like with React fast refresh), but I think for this to work with extension development the Directus instance would need to reload itself? Compared to the extension just rebuilding on code changes, I think the user would still need to reload the website manually, which isn't too bad of a DX but could still be better!

>>> Expose the directus app/api internals to custom interfaces
>>
>> This could be solved by starting to put things into a @directus/shared (name tbd) package that makes it a little easier to import those sorts of available internals, without having to rely on provide/inject for everything
>
> That would work! Seems like a good middle ground. The key point is being able to use the import x from @directus/shared to get IDE goodies. Then the bundling could be at runtime or buildtime.

`@directus/shared` would be a really good solution to getting the "official" types integrated into the toolkit too!

**Reply by rijkvanzanten** [reply]
> @directus/shared would be a really good solution to getting the "official" types integrated into the toolkit too!

💯 It's also frankly silly that the app, api, and SDK all have their own versions of the types, even though they're referring to the same objects..

**Comment by tylerforesthauser** [comment]
I've definitely found that Rollup can really make building custom interfaces a terrible experience, as I've had nothing but headaches trying to build a tiptap interface extension (while building it as a "native" interface works like a charm). Between having to bundle Vue in with your extension (if you use Typescript and/or @vue/composition-api) and having to fend for ourselves with regards to styling, the idea of _not_ bundling our extensions and having access to native interfaces/displays/etc makes me smile from ear to ear.

**Comment by cj** [comment]
It would be great to add a `--dev` flag to `directus start` that turns on hot reloading for any changes in the extensions folder.

**Reply by rijkvanzanten** [reply]
I'm wondering if Directus itself should restart fully, or just re-initialize the API extensions in that mode 🤔 The app extensions could hopefully be solved by hot reloading just the extension itself, instead of refreshing the whole app

**Reply by cj** [reply]
@rijkvanzanten thanks for this amazing project!

Yes - I would just have the extension hot reload.

**Comment by cj** [comment]
What is the best way currently in a custom component to access the current data of the collection being displayed, or to access another collection entirely?

**Reply by rijkvanzanten** [reply]
App or API?

**Reply by cj** [reply]
Both.

So right now I am writing a custom display, is it possible to access other collection data?

In a custom interface is it possible to do the same, but also access the current collection that is open?

**Reply by rijkvanzanten** [reply]
You can use https://docs.directus.io/guides/modules/#accessing-the-api-from-within-your-extension for interfaces/displays as well. Do keep in mind that displays might be rendered very often (eg once per column per row in the tabular layout), so I wouldn't recommend doing async data fetching on load of the display (only on interaction). The current collection is passed in the `collection` prop 👍🏻

**Reply by cj** [reply]
It looks like adding `inject: ['system']` which is in the https://docs.directus.io/guides/modules/ docs works.

**Reply by cj** [reply]
@rijkvanzanten thank you!

**Reply by rijkvanzanten** [reply]
Yup! 😄 I beat you to it by 7 seconds

**Reply by cj** [reply]
I think once HMR is added for custom extensions (when developing) Directus will be perfect!!!

**Reply by cj** [reply]
ha 😄

**Reply by rijkvanzanten** [reply]
Agreed! (cc @nickrum)

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
